tests/django_pg_migration_tools/test_operations.py

```diff
@@ -2701,3 +2702,347 @@ def test_operation_when_referred_model_is_defined_as_str(self):
+                operation.database_backwards(
+                    self.app_label, editor, project_state, new_state
+                )
+
+        # No queries have run, because the migration wasn't allowed to run by
+        # the router.
+        assert len(queries) == 0
+
+    @pytest.mark.django_db(transaction=True)
+    def test_basic_operation(self):
+        # Prove that the constraint does **not** already exist.
+        with connection.cursor() as cursor:
+            cursor.execute(
+                psycopg_sql.SQL(operations.ConstraintQueries.CHECK_EXISTING_CONSTRAINT)
+                .format(constraint_name=psycopg_sql.Literal("positive_int"))
+                .as_string(cursor.connection)
+            )
+            assert not cursor.fetchone()
+
+        project_state = ProjectState()
+        project_state.add_model(ModelState.from_model(IntModel))
+        new_state = project_state.clone()
+
+        operation = operations.SaferAddCheckConstraint(
+            model_name="intmodel",
+            constraint=self._get_check_constraint(
+                condition=Q(int_field__gte=0),
+                name="positive_int",
+            ),
+        )
+
+        assert operation.describe() == (
+            "Create constraint positive_int on model intmodel. "
+            "Note: Using django_pg_migration_tools SaferAddCheckConstraint "
+            "operation."
+        )
+
+        name, args, kwargs = operation.deconstruct()
+        assert name == "SaferAddCheckConstraint"
+        assert args == []
+        assert kwargs == {"model_name": "intmodel", "constraint": operation.constraint}
+
+        operation.state_forwards(self.app_label, new_state)
+        assert (
+            len(new_state.models[self.app_label, "intmodel"].options["constraints"])
+            == 1
+        )
+        assert (
+            new_state.models[self.app_label, "intmodel"].options["constraints"][0].name
+            == "positive_int"
+        )
+
+        with connection.schema_editor(atomic=False, collect_sql=False) as editor:
+            with utils.CaptureQueriesContext(connection) as queries:
+                operation.database_forwards(
+                    self.app_label, editor, project_state, new_state
+                )
+
+        assert len(queries) == 3
+
+        # 1. Check if the constraint is there.
+        assert queries[0]["sql"] == dedent("""
+            SELECT conname
+            FROM pg_catalog.pg_constraint
+            WHERE conname = 'positive_int';
+            """)
+
+        # 2. Add a not valid constraint
+        assert queries[1]["sql"] == (
+            'ALTER TABLE "example_app_intmodel" ADD CONSTRAINT "positive_int" '
+            'CHECK ("int_field" >= 0) NOT VALID;'
+        )
+
+        # 3. Validate it
+        assert queries[2]["sql"] == dedent("""
+            ALTER TABLE "example_app_intmodel"
+            VALIDATE CONSTRAINT "positive_int";
+        """)
+
```

Comment:
At this point it would be good to repeat the query from lines 2804-2809 along with an assertion that the constraint does now in fact exist. That would validate the creation of the constraint separately from the validation of the idempotency of the operation that's done next.

Reply:
Fixup commit: 13c02ebdc1d2fa9b47674c3d494f8da5fb6b8e4c